tests/test_conda_lock.py

```diff
@@ -2447,6 +2447,41 @@ def test_pip_finds_recent_manylinux_wheels(
     assert manylinux_version > [2, 17]
 
 
+def test_pip_respects_glibc_version(
+    tmp_path: Path, conda_exe: str, monkeypatch: "pytest.MonkeyPatch"
+):
+    """Ensure that we find a manylinux wheel that respects an older glibc constraint.
+
+    This is somewhat the opposite of test_pip_finds_recent_manylinux_wheels
+    """
+
+    env_file = clone_test_dir("test-pip-respects-glibc-version", tmp_path).joinpath(
+        "environment.yml"
+    )
+    monkeypatch.chdir(env_file.parent)
+    run_lock(
+        [env_file],
+        conda_exe=str(conda_exe),
+        platforms=["linux-64"],
+        virtual_package_spec=env_file.parent / "virtual-packages.yml",
+    )
+
+    lockfile = parse_conda_lock_file(env_file.parent / DEFAULT_LOCKFILE_NAME)
+
+    (cryptography_dep,) = [p for p in lockfile.package if p.name == "cryptography"]
+    manylinux_pattern = r"manylinux_(\d+)_(\d+).+\.whl"
+    # Should return the first match so higher version first.
+    manylinux_match = re.search(manylinux_pattern, cryptography_dep.url)
+    assert (
+        manylinux_match
+    ), "No match found for manylinux version in {cryptography_dep.url}"
+
+    manylinux_version = [int(each) for each in manylinux_match.groups()]
+    # Make sure the manylinux wheel was built with glibc <= 2.17
+    # since that is what the virtual package spec requires
+    assert manylinux_version <= [2, 17]
```

Comment:
I am asking it as a genuine question rather than criticism. Like semantically, if the `__glibc` has whatever version, does that mean that I can use wheels for anything <= that version, or does it need to be an exact match?

Reply:
Yes, my understanding is that GLIBC is backward compatible. See here for example: https://developers.redhat.com/blog/2019/08/01/how-the-gnu-c-library-handles-backward-compatibility.

Reply:
Thanks a lot for the explanation.

I don't understand the low-level details, but I think the following overview may be accurate:

There is some Conda magic to detect when a package was compiled using glibc. (I'm not sure exactly what magic, but glibc is only relevant for Linux platforms.) In this case it inserts a dependency that looks something like `__glibc >=2.17,<3.0.a0` in case glibc v2.17 was the version used, because later versions of glibc v2 can always run executables built with earlier versions.

An example package depending on glibc is [tzcode](https://github.com/conda-forge/tzcode-feedstock). The corresponding dependency in [the distributed file](https://anaconda.org/conda-forge/tzcode/2023c/download/linux-64/tzcode-2023c-h0b41bf4_0.conda) under `info-tzcode-2023c-h0b41bf4_0.tar.zst/info/index.json` is the `__glibc >=2.17,<3.0.a0` mentioned above.

Now glibc is a system-level dependency not managed by Conda. So when it comes time to install a Conda package on a Linux platform, Conda looks at the system's glibc version and adds a virtual package named `__glibc` to all environments with the corresponding system version. This way Conda's solver prevents installation of packages built with a version of glibc that's newer than the one installed on the system.

Finally, for pip-installed packages we need to determine which wheels are usable by conda-lock. We use fake virtual packages so that the assumed system-level dependencies are independent of the machine used to generate the lockfile. The strategy here is to determine the version of the fake `__glibc` virtual package. We then use this to filter the wheels.

Assuming that the fake `__glibc` version is 2.X and the wheel version is 2.Y, then the constraint `__glibc >=2.Y,<3.0.a0` is satisfied whenever X>=Y, so we filter for wheels with glibc version <=2.X. The manylinux spec of PEP 600 defines how to determine the glibc version from the tag.

Assuming this is correct, I wonder if we can include it, or perhaps a link to this comment?